component/main.jsonnet

```diff
@@ -30,7 +30,44 @@ local opaqueSecrets() = [
   for name in std.objectFields(params.opaque)
 ];
 
+local tlsSecrets() = [
+  local values = std.get(params.tls, name);
+  {
+    apiVersion: 'v1',
+    kind: 'Secret',
+    metadata: {
+      name: namespacedName(name, '').name,
+      namespace: namespacedName(name, '').namespace,
+    },
+    type: 'kubernetes.io/tls',
+    data: {
+      'tls.crt': values.key,
+      'tls.key': values.cert,
+    },
+  }
+  for name in std.objectFields(params.tls)
+];
```

Comment:
You should probably also support `ca.crt` here as an optional field. Additionally, you should probably handle invalid entries more explicitly. Right now a malformed entry for `tls` would just crash with a stack trace.

Also, you may need to append a trailing newline to the values to avoid hard to debug issues with software that can't deal with a PEM files that don't have a trailing newline. We already implemented this in component openshift4-ingress, see https://github.com/appuio/component-openshift4-ingress/blob/3d1173b7214b4ac704304ed7b51092895c9f4d9d/component/main.jsonnet#L30-L37 and https://github.com/appuio/component-openshift4-ingress/blob/3d1173b7214b4ac704304ed7b51092895c9f4d9d/component/main.jsonnet#L50-L59. See below for a rough variant which should work here.

Overall, I'd probably just expose the TLS secret keys directly, i.e. I'd make users specify TLS secrets as

```
tls:
  foo:
    tls.key: VALUE
    tls.crt: VALUE
```

That way, you don't need to map from the input keys to the correct secret keys (which you mixed up, as I noticed after staring at the implementation for 15 minutes).

Something like the following (untested):

```suggestion
local validTlsFields(values) =
  local secretKeys = std.set(std.objectFields(values));
  local keyDiff = std.setDiff(secretKeys, std.set([
    'ca.crt',
    'tls.crt',
    'tls.key',
  ]));
  keyDiff == 0;
  
local tlsSecrets() = [
  local values = {
    [k]: params.tls[name] + '\n'
    for k in std.objectFields(params.tls[name])
  };
  if validTlsFields(values) then
    {
      apiVersion: 'v1',
      kind: 'Secret',
      metadata: {
        name: namespacedName(name, '').name,
        namespace: namespacedName(name, '').namespace,
      },
      type: 'kubernetes.io/tls',
      stringData: values,
    }
  else
    error 'Invalid keys for TLS secret %s. Only keys `tls.key`, `tls.crt` and `ca.crt` are supported.' % name
  for name in std.objectFields(params.tls)
];
```

Reply:
Obsolete, as I have rewritten the logic completely.